dvc/external_repo.py

```diff
@@ -123,21 +123,32 @@ def _set_cache_dir(self):
 
         self.cache.local.cache_dir = cache_dir
 
-    def _set_upstream(self):
-        # check if the URL is local and no default remote is present
-        # add default remote pointing to the original repo's cache location
+    def _fix_upstream(self):
         if os.path.isdir(self.url):
-            if not self.config["core"].get("remote"):
-                src_repo = Repo(self.url)
-                try:
-                    cache_dir = src_repo.cache.local.cache_dir
-                finally:
-                    src_repo.close()
-
-                self.config["remote"]["auto-generated-upstream"] = {
-                    "url": cache_dir
-                }
-                self.config["core"]["remote"] = "auto-generated-upstream"
+            upstream_name = self.config["core"].get("remote")
+            src_repo = Repo(self.url)
+            try:
+                if upstream_name:
```

Comment:
```python
try:
   self._add_upstream(src_repo)
...
```

and move the logic of resolving upstream into the `_add_upstream` method. Cause here the upstream is added, what type of it and how - can be handled inside the method
If you want to use conditions here then methods can be named differently, for example as
```python
try:
    if remote_name:
      self._add_upstream_from_remote(..) # not sure in relevance of `add_upstream`
   else:
      self._add_upstream_from_repo(...)
```

Reply:
I'm thinking that maybe the two separate flows are not required? Consider my question here https://github.com/iterative/dvc/pull/3378#issuecomment-589820202

If instead of going one way or another depending on if there is a default remote set, I go through every remote and replace every RelPath with the one from the source repo, and then I check if there is a default repo set up (and handle that accordingly) there will be a single branch of execution.

Something like this:

```python
    def _fix_remotes(self):
        if not os.path.isdir(self.url):
            return

        src_repo = Repo(self.url)
        try:
            self._fix_source_relative_remotes(src_repo)
            if "remote" not in self.config["core"]:
                self._set_upstream(src_repo)
        finally:
            src_repo.close()

    def _fix_source_relative_remotes(self, src_repo):
        for remote_name, new_remote in self.config["remote"].items():
            if isinstance(new_remote["url"], RelPath):
                old_remote = src_repo.config["remote"][remote_name]
                new_remote["url"] = old_remote["url"]
```

/cc @efiop

Reply:
Maybe rename `_set_upstream` to `_set_upstream_to_repo` (also, the comment there is probably not needed anymore, it is describing what the three lines of code in that function already express clearly).

Reply:
@tizoc I think two branches of execution are necessary, since `_set_upstream` is only needed when there is no default remote and `_fix_upstream` is only needed when there is a default remote but it is a local one with relpath url.

@JIoJIaJIu So you mean moving
```
                if upstream_name:
                    self._fix_local_remote(src_repo, upstream_name)
                else:
                    self._add_upstream(src_repo)
```

into a new method? Well, we could, but I don't see much value in doing that, tbh. I would be fine with the current approach as well.

Reply:
@efiop my reasoning is, if relative urls are imported from the source repo, we want all of them to be kept relative to the source repo, right? not just the one that happens to be the default. If that is the case, that is independent from if there is a default remote set or not.

Reply:
@tizoc But we don't really care about other remotes in this case. We only care about default remote, as it is the only thing that we are actually going to use, at least for now.